pkg/catalogv2/oci/oci.go

```diff
@@ -0,0 +1,246 @@
+	manifestblob, err := content.FetchAll(context.Background(), memoryStore, manifest)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch the manifest blob of %s:%s", fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag), err.Error())
+	}
+
+	var manifestJSON ocispec.Manifest
+	err = json.Unmarshal(manifestblob, &manifestJSON)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal manifest blob of %s:%s", fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag), err.Error())
+	}
+
+	// Check if the oci artifact is of type helm config ?
+	if manifest.ArtifactType == helmregistry.ConfigMediaType || manifestJSON.Config.MediaType == helmregistry.ConfigMediaType {
+		// find the layer of chart type and fetch it
+		for _, layer := range manifestJSON.Layers {
+			if layer.MediaType == helmregistry.ChartLayerMediaType {
+
+				chartTar, err := content.FetchAll(context.Background(), memoryStore, layer)
+				if err != nil {
+					return nil, err
+				}
+
+				return io.NopCloser(bytes.NewBuffer(chartTar)), nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find the required chart tar file for %s", fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag))
+}
+
+// GenerateIndex creates an helm repo index from the oci url provided
+// by fetching the repositories or tags according to the url and
+// untaring the oci artifact chart and then adding the chart entry to the
+// helm repo index using the oras go library.
+func GenerateIndex(URL string, credentialSecret *corev1.Secret, clusterRepoSpec v1.RepoSpec) (*repo.IndexFile, error) {
+	logrus.Debugf("Generating index for oci clusterrepo URL %s", URL)
+
+	// Create a new oci client
+	ociClient, err := NewClient(URL, clusterRepoSpec, credentialSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an oci client for url %s: %s", URL, err.Error())
+	}
+
+	// Create an empty helm repo index file
+	indexFile := repo.NewIndexFile()
+
+	var orasRepository *remote.Repository
+	chartName := ""
+
+	// Loop over tags function
+	tagsFunc := func(tags []string) error {
+		for _, tag := range tags {
+			if _, err := semver.NewVersion(tag); err == nil {
+				logrus.Debugf("found a tag %s for the repository %s for OCI clusterrepo URL %s", tag, ociClient.repository, URL)
+
+				ociClient.tag = tag
+
+				// Fetch the helm chart
+				filePath, isHelmChart, err := ociClient.fetchChart(orasRepository)
+				if err != nil {
+					return fmt.Errorf("failed to fetch an helm chart %s: %s", fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag), err.Error())
+				}
+
+				// Add to the index
+				if isHelmChart {
+					err = ociClient.addToIndex(indexFile, chartName, filePath)
+					if err != nil {
+						return fmt.Errorf("unable to add helm chart %s to index: %s", fmt.Sprintf("%s/%s:%s", ociClient.registry, ociClient.repository, ociClient.tag), err.Error())
```

Comment:
We could simplify and wrap error:

```suggestion
						return fmt.Errorf("unable to add helm chart %s/%s:%s to index: %w", ociClient.registry, ociClient.repository, ociClient.tag, err)
```

Reply:
Refactored this part of the codebase.